src/System.Linq/src/System/Linq/DefaultIfEmpty.cs

```diff
@@ -17,24 +19,95 @@ public static IEnumerable<TSource> DefaultIfEmpty<TSource>(this IEnumerable<TSou
+            private readonly TSource _default;
+            private IEnumerator<TSource> _enumerator;
+
+            public DefaultIfEmptyIterator(IEnumerable<TSource> source, TSource defaultValue)
+            {
+                Debug.Assert(source != null);
+                _source = source;
+                _default = defaultValue;
+            }
+
+            public override Iterator<TSource> Clone()
+            {
+                return new DefaultIfEmptyIterator<TSource>(_source, _default);
+            }
+
+            public override bool MoveNext()
+            {
+                switch (state)
+                {
+                    case 1:
+                        _enumerator = _source.GetEnumerator();
+                        if (_enumerator.MoveNext())
+                        {
+                            current = _enumerator.Current;
+                            state = 2;
+                        }
+                        else
+                        {
+                            current = _default;
+                            state = -1;
+                        }
+                        return true;
+                    case 2:
+                        if (_enumerator.MoveNext())
+                        {
+                            current = _enumerator.Current;
+                            return true;
+                        }
+                        break;
+                }
+
+                Dispose();
+                return false;
+            }
+
+            public override void Dispose()
+            {
+                if (_enumerator != null)
+                {
+                    _enumerator.Dispose();
+                    _enumerator = null;
+                }
+
+                base.Dispose();
+            }
+
+            public TSource[] ToArray()
+            {
+                TSource[] array = _source.ToArray();
+                return array.Length == 0 ? new[] { _default } : array;
+            }
+
+            public List<TSource> ToList()
+            {
+                List<TSource> list = _source.ToList();
+                if (list.Count == 0)
+                    list.Add(_default);
+                return list;
+            }
+
+            public int GetCount(bool onlyIfCheap)
             {
-                if (e.MoveNext())
+                int count;
+                if (!onlyIfCheap || _source is ICollection<TSource> || _source is ICollection)
                 {
-                    do
-                    {
-                        yield return e.Current;
-                    } while (e.MoveNext());
+                    count = _source.Count();
                 }
                 else
                 {
-                    yield return defaultValue;
+                    IIListProvider<TSource> listProv = _source as IIListProvider<TSource>;
+                    count = listProv == null ? -1 : listProv.GetCount(onlyIfCheap: true);
```

Comment:
@stephentoub I altered this since you said "LGTM". The only change was to change `GetCount(onlyIfCheap)` to `GetCount(onlyIfCheap: true)`. Since this path is only hit if that value is true this is no change, but it seemed clearer than passing a value that has just been checked for a particular value, to just pass that value.